File: WeeklyCode/week1/Search in Rotated Sorted Array II/chenhao.cpp
class Solution {
public:
    bool search(int A[], int n, int target) {
        int offset=0;
        for(int i=1;i<n;i++){
        	if(A[i]<A[i-1]){
        		offset=i;
        		break;
        	}
        }

        int lt=0,rt=n-1;
        while(lt<=rt){
        	int mid=(lt+rt)>>1;
        	int value=A[(mid+offset)%n];
        	if(value==target){
        		return true;
        	}
        	else if(value<target){
        		lt=mid+1;
        	}
        	else{
        		rt=mid-1;
        	}
        }
        return false;
    }
};

File: WeeklyCode/week11/Search a 2D Matrix/chenhao.java
public class Solution {

    public boolean searchMatrix(int[][] matrix, int target) {
        int n=0,m=0;
        if( (n=matrix.length)==0 || (m=matrix[0].length)==0 ){
            return false;
        }
        int lo=0, hi=n*m-1;
        while(lo<=hi){
            int mid=(lo+hi)/2;
            int val=matrix[mid/m][mid%m];
            if(val==target){
                return true;
            }
            else if(val>target){
                hi=mid-1;
            }
            else{
                lo=mid+1;
            }
        }
        return false;
    }
}